Add customized document serialization support

remcolam opened the pull request:
Added support to inject a ISwaggerDocumentSerializer implementation, which can be called to manually serialize an OpenAPIDocument object.

- Resolves #2668
- Resolves #2700
- Resolves #2105

martincostello commented:
Thanks for contributing - if you'd like to continue with this pull request, please rebase against the default branch to pick up our new CI.

Tests will also be required.

martincostello commented on src/Swashbuckle.AspNetCore.Swagger/ISwaggerDocumentSerializer.cs:
```diff
@@ -0,0 +1,11 @@
+
+namespace Swashbuckle.AspNetCore.Swagger
+{
+    public interface ISwaggerDocumentSerializer
+    {
+		void SerializeDocument(OpenApiDocument document, IOpenApiWriter writer, OpenApiSpecVersion specVersion);
+	}
```
Please add `///` comments to the interface and the method.

martincostello commented on src/Swashbuckle.AspNetCore.Swagger/SwaggerMiddleware.cs:
```diff
@@ -15,14 +16,19 @@ public class SwaggerMiddleware
         public SwaggerMiddleware(
             RequestDelegate next,
-            SwaggerOptions options)
+            SwaggerOptions options,
+            IServiceProvider serviceProvider)
```
This is a binary-breaking change, so we could only take this as part of a major release (i.e. v7).

remcolam replied:
Would it also be breaking if we add it as a second constructor? And linking the previous constructor to this one?

martincostello replied:
No that wouldn't be breaking, provided that DI knows which constructor to use.

remcolam replied:
Quick question, should I "resolve" conversations myself when I've handled them (even if I didn't push yet) or leave that to the original commenter?

martincostello commented on src/Swashbuckle.AspNetCore.Swagger/SwaggerMiddleware.cs:
```diff
@@ -15,14 +16,19 @@ public class SwaggerMiddleware
             _requestMatcher = new TemplateMatcher(TemplateParser.Parse(_options.RouteTemplate), new RouteValueDictionary());
+
+            // Use IServiceProvider to retrieve the ISwaggerDocumentSerializer, because it is an optional service
+            _swaggerDocumentSerializer = serviceProvider.GetService(typeof(ISwaggerDocumentSerializer)) as ISwaggerDocumentSerializer;
```
Use `GetService<ISwaggerDocumentSerializer>()`?

martincostello commented on src/Swashbuckle.AspNetCore.SwaggerGen/DependencyInjection/DocumentProvider.cs:
```diff
@@ -24,15 +25,21 @@ internal class DocumentProvider : IDocumentProvider
             _swaggerProvider = swaggerProvider;
+
+            // Use IServiceProvider to retrieve the ISwaggerDocumentSerializer, because it is an optional service
+            _swaggerDocumentSerializer = serviceProvider.GetService(typeof(ISwaggerDocumentSerializer)) as ISwaggerDocumentSerializer;
```
```suggestion
            _swaggerDocumentSerializer = serviceProvider.GetService<ISwaggerDocumentSerializer>();
```

remcolam commented:
I will have to look into the tests still, I always struggle with creating those, maybe I'll ask for some help from some colleagues

martincostello commented on src/Swashbuckle.AspNetCore.Swagger/ISwaggerDocumentSerializer.cs:
```diff
@@ -0,0 +1,21 @@
+namespace Swashbuckle.AspNetCore.Swagger
+{
+    /// <summary>
+    /// Provide an implementation for this interface if you wish to customize how the open api document is exactly written.
```
```suggestion
    /// Provide an implementation for this interface if you wish to customize how the OpenAPI document is written.
```

martincostello commented on src/Swashbuckle.AspNetCore.Swagger/ISwaggerDocumentSerializer.cs:
```diff
@@ -0,0 +1,21 @@
+{
+    /// <summary>
+    /// Provide an implementation for this interface if you wish to customize how the open api document is exactly written.
+    /// SerializeDocument will be called in that case, instead of the Microsoft built in Serialize methods.
```
```suggestion
```

remcolam replied:
I might be going blind, but what exactly is different in this suggestion?

remcolam replied:
🤦‍♂️

martincostello commented on src/Swashbuckle.AspNetCore.Swagger/ISwaggerDocumentSerializer.cs:
```diff
@@ -0,0 +1,21 @@
+    /// </summary>
+    public interface ISwaggerDocumentSerializer
+    {
+        /// <summary>
+        /// Called in the places where normally SerializeV2 or SerializeV3 on OpenApiDocument would be called.
+        /// </summary>
+        /// <param name="document">The open api document that should be serialized</param>
+        /// <param name="writer">The write to which the document needs to be written</param>
+        /// <param name="specVersion">The open api spec to serialize</param>
```
```suggestion
        /// <summary>
        /// Serializes an OpenAPI document.
        /// </summary>
        /// <param name="document">The OpenAPI document that should be serialized.</param>
        /// <param name="writer">The writer to which the document needs to be written.</param>
        /// <param name="specVersion">The OpenAPI specification version to serialize as.</param>
```

martincostello commented on src/Swashbuckle.AspNetCore.Swagger/SwaggerMiddleware.cs:
```diff
@@ -15,14 +17,25 @@ public class SwaggerMiddleware
         public SwaggerMiddleware(
             RequestDelegate next,
-            SwaggerOptions options)
+            SwaggerOptions options) : this (next, options, null)
```
```suggestion
            SwaggerOptions options) : this(next, options, null)
```

martincostello commented on src/Swashbuckle.AspNetCore.Swagger/SwaggerMiddleware.cs:
```diff
@@ -15,14 +17,25 @@ public class SwaggerMiddleware
             _next = next;
             _options = options ?? new SwaggerOptions();
             _requestMatcher = new TemplateMatcher(TemplateParser.Parse(_options.RouteTemplate), new RouteValueDictionary());
+
+            // Use IServiceProvider to retrieve the ISwaggerDocumentSerializer, because it is an optional service
+            _swaggerDocumentSerializer = serviceProvider?.GetService<ISwaggerDocumentSerializer>();
```
I think this is self-documenting.

```suggestion
            _swaggerDocumentSerializer = serviceProvider?.GetService<ISwaggerDocumentSerializer>();
```

remcolam replied:
I am wondering though... I have now added IServiceProvider in an overloaded constructor. I could also directly add ISwaggerDocumentSerializer. 

For the current setup, that would be fine, but not if in the future more optional interfaces are added. Then you would need overloads for each possible combination of available interfaces.

What is your preference?

martincostello commented on src/Swashbuckle.AspNetCore.SwaggerGen/DependencyInjection/DocumentProvider.cs:
```diff
@@ -24,15 +26,28 @@ internal class DocumentProvider : IDocumentProvider
             IOptions<SwaggerGeneratorOptions> generatorOptions,
             IOptions<SwaggerOptions> options,
-            IAsyncSwaggerProvider swaggerProvider)
+            IAsyncSwaggerProvider swaggerProvider
+            ) : this(generatorOptions, options, swaggerProvider, null)
```
```suggestion
            IAsyncSwaggerProvider swaggerProvider)
            : this(generatorOptions, options, swaggerProvider, null)
```

remcolam replied:
If I put the closing bracket on this location here, I think it should also go to that location in the other constructor

martincostello commented on src/Swashbuckle.AspNetCore.SwaggerGen/DependencyInjection/DocumentProvider.cs:
```diff
@@ -24,15 +26,28 @@ internal class DocumentProvider : IDocumentProvider
+            IOptions<SwaggerGeneratorOptions> generatorOptions,
+            IOptions<SwaggerOptions> options,
+            IAsyncSwaggerProvider swaggerProvider,
+            IServiceProvider serviceProvider
+            )
```
```suggestion
            IServiceProvider serviceProvider)
```

martincostello commented on src/Swashbuckle.AspNetCore.SwaggerGen/DependencyInjection/DocumentProvider.cs:
```diff
@@ -24,15 +26,28 @@ internal class DocumentProvider : IDocumentProvider
             _generatorOptions = generatorOptions.Value;
             _options = options.Value;
             _swaggerProvider = swaggerProvider;
+
+            // Use IServiceProvider to retrieve the ISwaggerDocumentSerializer, because it is an optional service
+            _swaggerDocumentSerializer = serviceProvider?.GetService<ISwaggerDocumentSerializer>();
```
```suggestion
            _swaggerDocumentSerializer = serviceProvider?.GetService<ISwaggerDocumentSerializer>();
```

martincostello reviewed (commented):
Looks OK to me so far, modulo a few style comments.

martincostello commented on src/Swashbuckle.AspNetCore.Swagger/SwaggerMiddleware.cs:
```diff
@@ -15,14 +17,24 @@ public class SwaggerMiddleware
             _next = next;
             _options = options ?? new SwaggerOptions();
             _requestMatcher = new TemplateMatcher(TemplateParser.Parse(_options.RouteTemplate), new RouteValueDictionary());
+
```
Minor nit: I don't think there's a need for this assignment to be separated from the other three.

```suggestion
```

remcolam replied:
Yeah, I think it is leftover from removing the comment... when the comment was there it looked better with a whitespace...

martincostello commented on src/Swashbuckle.AspNetCore.SwaggerGen/DependencyInjection/DocumentProvider.cs:
```diff
@@ -24,15 +26,26 @@ internal class DocumentProvider : IDocumentProvider
             _generatorOptions = generatorOptions.Value;
             _options = options.Value;
             _swaggerProvider = swaggerProvider;
+
```
```suggestion
```

remcolam commented:
@martincostello As I am adding a test, I have the CLI test failing for me. Swashbuckle.AspNetCore.Cli.Test.ToolTests.Can_Generate_Swagger_Json() fails, because my local culture is serializing the OpenApiDouble(14.37) price example with a comma, not a period. 

It seems the CLI is not correctly serializing different cultures. Should I create an issue for this?

(FYI, I have actually run into this myself when adding examples, and solved it by creating a SwaggerHostFactory and specifically setting the invariant culture in there)

martincostello commented:
> Should I create an issue for this?

I believe that's #2105 and #2700, which there's a proposed fix for in #2726.

The author of that PR hasn't rebased, so I can't merge it. If you pull that change into your PR and it fixes it, then we can include that fix here too.

martincostello commented on Swashbuckle.AspNetCore.sln:
```diff
@@ -501,6 +503,18 @@ Global
 		{3D0126CB-5439-483C-B2D5-4B4BE111D15C}.Release|x64.Build.0 = Release|Any CPU
 		{3D0126CB-5439-483C-B2D5-4B4BE111D15C}.Release|x86.ActiveCfg = Release|Any CPU
 		{3D0126CB-5439-483C-B2D5-4B4BE111D15C}.Release|x86.Build.0 = Release|Any CPU
+		{B6037A37-4A4F-438D-B18A-0C9D1408EAB2}.Debug|Any CPU.ActiveCfg = Debug|Any CPU
+		{B6037A37-4A4F-438D-B18A-0C9D1408EAB2}.Debug|Any CPU.Build.0 = Debug|Any CPU
+		{B6037A37-4A4F-438D-B18A-0C9D1408EAB2}.Debug|x64.ActiveCfg = Debug|Any CPU
+		{B6037A37-4A4F-438D-B18A-0C9D1408EAB2}.Debug|x64.Build.0 = Debug|Any CPU
+		{B6037A37-4A4F-438D-B18A-0C9D1408EAB2}.Debug|x86.ActiveCfg = Debug|Any CPU
+		{B6037A37-4A4F-438D-B18A-0C9D1408EAB2}.Debug|x86.Build.0 = Debug|Any CPU
+		{B6037A37-4A4F-438D-B18A-0C9D1408EAB2}.Release|Any CPU.ActiveCfg = Release|Any CPU
+		{B6037A37-4A4F-438D-B18A-0C9D1408EAB2}.Release|Any CPU.Build.0 = Release|Any CPU
+		{B6037A37-4A4F-438D-B18A-0C9D1408EAB2}.Release|x64.ActiveCfg = Release|Any CPU
+		{B6037A37-4A4F-438D-B18A-0C9D1408EAB2}.Release|x64.Build.0 = Release|Any CPU
+		{B6037A37-4A4F-438D-B18A-0C9D1408EAB2}.Release|x86.ActiveCfg = Release|Any CPU
+		{B6037A37-4A4F-438D-B18A-0C9D1408EAB2}.Release|x86.Build.0 = Release|Any CPU
```
We only need the Any CPU configuration here.

martincostello commented on src/Swashbuckle.AspNetCore.Cli/Program.cs:
```diff
@@ -10,6 +10,7 @@
 using Microsoft.AspNetCore.Hosting;
 using Microsoft.AspNetCore;
 using Microsoft.Extensions.Hosting;
+using System.Globalization;
```
Nit: sort using.

martincostello commented on src/Swashbuckle.AspNetCore.Swagger/SwaggerMiddleware.cs:
```diff
@@ -33,10 +27,16 @@ public class SwaggerMiddleware
             _swaggerDocumentSerializer = serviceProvider?.GetService<ISwaggerDocumentSerializer>();
         }
 
+        [Obsolete("Remove in next major release, kept for binairy compatibility")]
```
Just add this as a `// TODO` comment. Otherwise things that use `TreatWarningsAsErrors=true` will break.

I've added a comment to track this here: https://github.com/domaindrivendev/Swashbuckle.AspNetCore/issues/2798#issuecomment-2060853358

remcolam commented:
I had trouble finding good tests, I hope, the 6 I added are satisfactory.

I added the fix for the culture in CLI, copied the change from the other PR.

I marked the original constructors that I overloaded obsolete, I think they should be removed in the next major release, there shouldn't be multiple injection constructors, and changing the order of them breaks the DI also. (Which my test found out)

martincostello commented on src/Swashbuckle.AspNetCore.SwaggerGen/DependencyInjection/DocumentProvider.cs:
```diff
@@ -44,10 +37,17 @@ internal class DocumentProvider : IDocumentProvider
             _swaggerDocumentSerializer = serviceProvider?.GetService<ISwaggerDocumentSerializer>();
         }
 
+        [Obsolete("Remove in next major release, kept for binairy compatibility")]
+        public DocumentProvider(
+            IOptions<SwaggerGeneratorOptions> generatorOptions,
+            IOptions<SwaggerOptions> options,
+            IAsyncSwaggerProvider swaggerProvider)
+            : this(generatorOptions, options, swaggerProvider, null)
+        { }
```
Same here.

martincostello commented on test/Swashbuckle.AspNetCore.Cli.Test/ToolTests.cs:
```diff
@@ -21,7 +21,7 @@ public void Can_Generate_Swagger_Json()
                 var args = new string[] { "tofile", "--output", $"{dir}/swagger.json", "--serializeasv2", Path.Combine(Directory.GetCurrentDirectory(), "Basic.dll"), "v1" };
                 Assert.Equal(0, Program.Main(args));
-
+                var x = File.ReadAllText(Path.Combine(dir.FullName, "swagger.json"));
```
Debugging leftovers?

martincostello commented on test/Swashbuckle.AspNetCore.Cli.Test/ToolTests.cs:
```diff
@@ -35,6 +35,50 @@ public void Can_Generate_Swagger_Json()
+                dir.Delete(true);
+            }
+        }
+
+
+        [Fact]
```
```suggestion

        [Fact]
```

martincostello commented on test/Swashbuckle.AspNetCore.Cli.Test/ToolTests.cs:
```diff
@@ -35,6 +35,50 @@ public void Can_Generate_Swagger_Json()
+            }
+            finally
+            {
+                dir.Delete(true);
```
Might be useful to "borrow" this utility class I wrote recently and add it to the test helpers to use in these tests: [TemporaryDirectory](https://github.com/martincostello/dotnet-bumper/blob/main/src/DotNetBumper.Core/TemporaryDirectory.cs)

remcolam replied:
Added a version of it to Swashbuckle.AspNetCore.TestSupport

martincostello commented on test/Swashbuckle.AspNetCore.IntegrationTests/CustomDocumentSerializerTests.cs:
```diff
@@ -0,0 +1,52 @@
+using Microsoft.OpenApi.Any;
+using Microsoft.OpenApi.Readers;
+using Swashbuckle.AspNetCore.Swagger;
+using Xunit;
```
This is a lot of usings - I feel some of them are redundant and can be deleted.

remcolam replied:
I did a sort + clean on them

martincostello commented on test/WebSites/CustomDocumentSerializer/DocumentSerializerTest.cs:
```diff
@@ -0,0 +1,27 @@
+                writer.WriteProperty(OpenApiConstants.Swagger, "DocumentSerializerTest3.0");
+                break;
+            default:
+                throw new ArgumentOutOfRangeException(nameof(specVersion), $"Unexpected value '{specVersion}'");
```
As this is just test code:

```suggestion
                throw new NotImplementedException();
```

martincostello commented on test/WebSites/CustomDocumentSerializer/Program.cs:
```diff
@@ -0,0 +1,22 @@
+
+namespace CustomDocumentSerializer
+{
+
```
```suggestion
```

martincostello commented on test/WebSites/CustomDocumentSerializer/Startup.cs:
```diff
@@ -0,0 +1,48 @@
+
+    public IConfiguration Configuration { get; }
+
+    // This method gets called by the runtime. Use this method to add services to the container.
```
```suggestion
```

martincostello commented on test/WebSites/CustomDocumentSerializer/Startup.cs:
```diff
@@ -0,0 +1,48 @@
+        });
+    }
+
+    // This method gets called by the runtime. Use this method to configure the HTTP request pipeline.
```
```suggestion
```

martincostello commented on test/WebSites/CustomDocumentSerializer/Startup.cs:
```diff
@@ -0,0 +1,48 @@
+    // This method gets called by the runtime. Use this method to add services to the container.
+    public void ConfigureServices(IServiceCollection services)
+    {
+        // Add services to the container.
```
```suggestion
```

martincostello commented on test/WebSites/CustomDocumentSerializer/Startup.cs:
```diff
@@ -0,0 +1,48 @@
+    // This method gets called by the runtime. Use this method to configure the HTTP request pipeline.
+    public void Configure(IApplicationBuilder app, IWebHostEnvironment env)
+    {
+        // Configure the HTTP request pipeline.
```
```suggestion
```

martincostello reviewed (commented):
Just a few minor comments.

martincostello commented:
Thanks for adding the tests too!

martincostello commented on src/Swashbuckle.AspNetCore.SwaggerGen/DependencyInjection/DocumentProvider.cs:
```diff
@@ -40,7 +40,7 @@ internal class DocumentProvider : IDocumentProvider
         }
 
-        [Obsolete("Remove in next major release, kept for binairy compatibility")]
+        // TODO: Remove in next major release, kept for binairy compatibility
```
```suggestion
        // TODO: Remove in next major release, kept for binary compatibility
```

remcolam commented on src/Swashbuckle.AspNetCore.Swagger/SwaggerMiddleware.cs:
```diff
@@ -30,7 +30,7 @@ public class SwaggerMiddleware
         }
 
-        [Obsolete("Remove in next major release, kept for binairy compatibility")]
+        // TODO: Remove in next major release, kept for binairy compatibility
```
Oops, I wrote it in Dunglish 😳

martincostello commented:
Thanks again for your work on this - I'll merge this once/when the next planned release is intended to be 6.6.0.

remcolam commented:
@martincostello I was looking at what I am trying to achieve again. I am wondering if I shouldn't change the place I am hooking in.

Instead of using DI, and registering the serializer in the service provider, it could be added to SwaggerGeneratorOptions. It will prevent changing the constructor of DocumentProvider and SwaggerMiddleware. And since it's an optional serializer, it shouldn't really go via DI.

But if you think, never mind, this is good enough also for other future use, that's fine

martincostello commented:
Maybe have a quick look at doing that locally in your branch and see if that makes things simpler or not? My only thought is that it _might_ make it more difficult for people to assign a custom `ISwaggerDocumentSerializer` in the case that they need to create it from other services registered in DI. If that's still possible with adding it to `SwaggerOptions` and the isn't a snowball effect of needing to update other things then that seems reasonable.

remcolam commented:
Who is responsible for solving the conflicts?

I noticed this in the conflicts... 

![image](https://github.com/domaindrivendev/Swashbuckle.AspNetCore/assets/11367055/17e421f7-4489-4cf9-aeb9-5736f464cf45)

During my testing I noticed, DI picks the top constructor. my new constructor and the added one need to be merged into a single one, with the new parameter encapsulated by `#if !NETSTANDARD`

martincostello commented:
The short answer is: you 😄

Regarding the constructor, that can be solved by things such as [`[ActivatorUtilitiesConstructor]`](https://learn.microsoft.com/dotnet/api/microsoft.extensions.dependencyinjection.activatorutilitiesconstructorattribute) to give a hint to the DI system, or changing the DI registration to explicitly pick a specific constructor.

remcolam commented:
Fixed merge conflicts. @martincostello Take a close look at the SwaggerMiddleware.cs, where I merged the new changes from main

martincostello commented on src/Swashbuckle.AspNetCore.Swagger/SwaggerMiddleware.cs:
The changes in this file seem fine after the rebase, but this code needs updating for the additional parameter:

https://github.com/domaindrivendev/Swashbuckle.AspNetCore/blob/a920aebc1456043c4f86e0b47432307176e86247/src/Swashbuckle.AspNetCore.Swagger/DependencyInjection/SwaggerBuilderExtensions.cs#L23-L27

remcolam replied:
Thanks for pointing it out. 

This convinces me more that I need to put the Custom document serializer into the SwaggerGenOptions (can't be in the more appropriate SwaggerOptions, because that only gets injected when running the application (in the Application)).

I think there is to opportunities for the constructor to get called without this value being passed in.

Agreed?

remcolam replied:
(FYI, the reason it can not be in SwaggerOptions is because that is not available in the CLI executable, since that is only configuring the host, but not running it)

martincostello replied:
I'm not sure why the current design is layered/architected in the way it is in terms of the layers.

remcolam commented:
@martincostello I abondened my idea of putting it into any of the options for obvious reasons, I fixed the call in SwaggerBuilderExtensions.cs, so I think now it should all be good

martincostello commented on src/Swashbuckle.AspNetCore.Swagger/DependencyInjection/SwaggerBuilderExtensions.cs:
```diff
@@ -21,9 +21,9 @@ public static class SwaggerBuilderExtensions
         {
 #if (!NETSTANDARD2_0)
-            return app.UseMiddleware<SwaggerMiddleware>(options, app.ApplicationServices.GetRequiredService<TemplateBinderFactory>());
+            return app.UseMiddleware<SwaggerMiddleware>(options, app.ApplicationServices.GetRequiredService<TemplateBinderFactory>(), app.ApplicationServices.GetService<ISwaggerDocumentSerializer>());
 #else
-            return app.UseMiddleware<SwaggerMiddleware>(options);
+            return app.UseMiddleware<SwaggerMiddleware>(options, app.ApplicationServices.GetService<ISwaggerDocumentSerializer>());
 #endif
```
So looking at this code vs the constructor, this should be passing in `app.ApplicationServices` as the last parameter, rather than the first.

_However_ that is highlighting that the other change made recently (why you need to rebase) gets that out of the service provider _anyway_, so instead we could try simplifying this like so:

```csharp
public static IApplicationBuilder UseSwagger(this IApplicationBuilder app, SwaggerOptions options)
{
    return app.UseMiddleware<SwaggerMiddleware>(options, app.ApplicationServices);
}

public SwaggerMiddleware(
    RequestDelegate next,
    SwaggerOptions options,
    IServiceProvider serviceProvider)
{
    _next = next;
    _options = options ?? new SwaggerOptions();
    _requestMatcher = new TemplateMatcher(TemplateParser.Parse(_options.RouteTemplate), new RouteValueDictionary());
    _swaggerDocumentSerializer = serviceProvider?.GetService<ISwaggerDocumentSerializer>();
#if !NETSTANDARD
    _templateBinder = serviceProvider?.GetService<TemplateBinderFactory>()?.Create(RoutePatternFactory.Parse(_options.RouteTemplate));
#endif
}

public SwaggerMiddleware(RequestDelegate next, SwaggerOptions options)
    : this(next, options, null)
{
}
```

remcolam replied:
Good point, done that now

remcolam commented:
@martincostello I switched it to using swagger options. I can of course revert back to the previous commit, so where it uses DI, but everywhere I read, they say optional DI is an anti pattern....

martincostello commented on src/Swashbuckle.AspNetCore.SwaggerGen/DependencyInjection/SwaggerGenOptions.cs:
```diff
@@ -1,5 +1,6 @@
 ﻿using System;
 using System.Collections.Generic;
+using Swashbuckle.AspNetCore.Swagger;
```
```suggestion
```

martincostello commented on src/Swashbuckle.AspNetCore.SwaggerGen/DependencyInjection/SwaggerGenOptionsExtensions.cs:
```diff
@@ -3,6 +3,7 @@
 using System.Xml.XPath;
 using Microsoft.OpenApi.Models;
 using Microsoft.AspNetCore.Mvc.ApiExplorer;
+using Swashbuckle.AspNetCore.Swagger;
```
```suggestion
```

martincostello commented on test/WebSites/CustomDocumentSerializer/Startup.cs:
```diff
@@ -0,0 +1,44 @@
+        {
+            endpoints.MapControllers();
+            endpoints.MapSwagger("swagger/{documentName}/swagger.json");
+            endpoints.MapSwagger("swagger/{documentName}/swaggerv2.json", c =>
+            {
+                c.SerializeAsV2 = true;
+            });
```
```suggestion
            endpoints.MapSwagger("swagger/{documentName}/swaggerv2.json", c => c.SerializeAsV2 = true);
```

martincostello commented on src/Swashbuckle.AspNetCore.Swagger/SwaggerOptions.cs:
```diff
@@ -25,9 +25,15 @@ public SwaggerOptions()
         /// </summary>
         public bool SerializeAsV2 { get; set; }
 
+        /// <summary>
+        /// Provide a custom ISwaggerDocumentSerializer implementation, to have more control over how the swagger document is exactly serialized.
+        /// </summary>
+        /// <remarks>For the CLI tool to be able to use this, this needs to be set during ConfigureServices.</remarks>
+        public ISwaggerDocumentSerializer CustomDocumentSerializer { get; set; } = null;
```
Tweaked the wording here, as there are other ways to configure an ASP.NET Core application's dependencies other than using `ConfigureServices()`.

```suggestion
        /// <summary>
        /// Gets or sets an optional custom <see cref="ISwaggerDocumentSerializer"/> implementation to use to serialize Swagger documents.
        /// </summary>
        /// <remarks>For the CLI tool to be able to use this, this needs to be configured for use in the service collection of your application.</remarks>
        public ISwaggerDocumentSerializer CustomDocumentSerializer { get; set; }
```

martincostello commented on src/Swashbuckle.AspNetCore.Swagger/DependencyInjection/SwaggerServiceCollectionExtensions.cs:
```diff
@@ -0,0 +1,14 @@
+
+public static class SwaggerServiceCollectionExtensions
+{
+    public static void ConfigureSwagger(
```
Could you add `///` docs please?

martincostello commented on src/Swashbuckle.AspNetCore.Swagger/DependencyInjection/SwaggerOptionsExtensions.cs:
```diff
@@ -0,0 +1,21 @@
+
+public static class SwaggerOptionsExtensions
+{
+    /// <summary>
+    /// Sets a custom swagger document serializer.
+    /// </summary>
+    /// <remarks>To work for the CLI tool, this needs to be performed during ConfigureServices.</remarks>
+    /// <param name="swaggerOptions"></param>
+    /// <param name="constructorParameters">parameters to pass into the constructor of the custom swagger document serializer.</param>
```
```suggestion
    /// <summary>
    /// Sets a custom Swagger document serializer to use.
    /// </summary>
    /// <remarks>For the CLI tool to be able to use this, this needs to be configured for use in the service collection of your application.</remarks>
    /// <param name="swaggerOptions">The options to configure for serializer for.</param>
    /// <param name="constructorParameters">Theparameters to pass into the constructor of the custom Swagger document serializer implementation.</param>
```

martincostello commented on src/Swashbuckle.AspNetCore.Swagger/DependencyInjection/SwaggerOptionsExtensions.cs:
```diff
@@ -0,0 +1,21 @@
+        params object[] constructorParameters)
+        where TDocumentSerializer : ISwaggerDocumentSerializer
+    {
+        swaggerOptions.CustomDocumentSerializer = (TDocumentSerializer)Activator.CreateInstance(typeof(TDocumentSerializer), constructorParameters);
```
Check `swaggerOptions` for `null`.

remcolam replied:
Did it, but be aware it is also not done in any of the SwaggerGenOptionsExtensions

martincostello replied:
The danger of a new person reviewing code they've not seen before and didn't originally write - they'll find new things 😄

martincostello commented on src/Swashbuckle.AspNetCore.Swagger/DependencyInjection/SwaggerOptionsExtensions.cs:
```diff
@@ -0,0 +1,21 @@
+    /// <remarks>To work for the CLI tool, this needs to be performed during ConfigureServices.</remarks>
+    /// <param name="swaggerOptions"></param>
+    /// <param name="constructorParameters">parameters to pass into the constructor of the custom swagger document serializer.</param>
+    public static void SetCustomDocumentSerializer<TDocumentSerializer>(
```
/// docs are missing for the type parameter.

martincostello commented on src/Swashbuckle.AspNetCore.Swagger/DependencyInjection/SwaggerOptionsExtensions.cs:
```diff
@@ -0,0 +1,21 @@
+public static class SwaggerOptionsExtensions
+{
+    /// <summary>
+    /// Sets a custom swagger document serializer.
```
```suggestion
    /// Sets a custom Swagger document serializer.
```

martincostello commented:
Minus feedback, do you think you're done now @remcolam? I was poised to merge this when the last set of changes came in 😄

remcolam commented:
@martincostello If you are happy, then I am happy. I prefer going the route of the swagger options, it feels cleaner. I think I addressed all your last comments

martincostello commented on src/Swashbuckle.AspNetCore.Swagger/DependencyInjection/SwaggerOptionsExtensions.cs:
```diff
@@ -0,0 +1,29 @@
+namespace Microsoft.Extensions.DependencyInjection;
+
+/// <summary>
+/// Extensions for helping with configuring the SwaggerOptions
```
```suggestion
/// Extensions for helping with configuring instances of <see cref="SwaggerOptions"/>.
```

martincostello commented on src/Swashbuckle.AspNetCore.Swagger/DependencyInjection/SwaggerOptionsExtensions.cs:
```diff
@@ -0,0 +1,29 @@
+    /// Sets a custom Swagger document serializer to use.
+    /// </summary>
+    /// <remarks>For the CLI tool to be able to use this, this needs to be configured for use in the service collection of your application.</remarks>
+    /// <typeparam name="TDocumentSerializer">The custom Swagger document serializer implementation.</typeparam>
```
```suggestion
    /// <typeparam name="TDocumentSerializer">The type of the custom Swagger document serializer implementation.</typeparam>
```

martincostello commented on src/Swashbuckle.AspNetCore.Swagger/DependencyInjection/SwaggerOptionsExtensions.cs:
```diff
@@ -0,0 +1,29 @@
+    {
+        if (swaggerOptions == null)
+        {
+            return;
```
This draws the caller to their error, rather than silently hiding it.

```suggestion
            throw new ArgumentNullException(nameof(swaggerOptions));
```

martincostello commented on src/Swashbuckle.AspNetCore.Swagger/DependencyInjection/SwaggerServiceCollectionExtensions.cs:
```diff
@@ -0,0 +1,22 @@
+
+namespace Microsoft.Extensions.DependencyInjection;
+
+/// <summary>
+/// Extensions to configure swagger in the service collection
+/// </summary>
+public static class SwaggerServiceCollectionExtensions
+{
+    /// <summary>
+    /// Configure swagger options in the service collection.
+    /// </summary>
+    /// <param name="services">The service collection to set the swagger options on.</param>
+    /// <param name="setupAction">The setup action for swagger.</param>
```
```suggestion
/// <summary>
/// Extensions to configure dependencies for Swagger.
/// </summary>
public static class SwaggerServiceCollectionExtensions
{
    /// <summary>
    /// Configures Swagger options in the specified service collection.
    /// </summary>
    /// <param name="services">The service collection to configure the Swagger options for.</param>
    /// <param name="setupAction">A delegate to a method to use to configure the Swagger options.</param>
```

martincostello reviewed (commented):
Just a few last nit-picks then this should be good to merge 👍

remcolam commented:
Accepted all your changes

martincostello commented on src/Swashbuckle.AspNetCore.SwaggerGen/SwaggerGenerator/SwaggerGeneratorOptions.cs:
```diff
@@ -6,7 +6,6 @@
 #endif
 using Microsoft.AspNetCore.Mvc.ApiExplorer;
 using Microsoft.OpenApi.Models;
-using Microsoft.AspNetCore.Routing;
```
Missed this - needs putting back as it breaks the build.

remcolam replied:
Shouldn't it be in the compiler switch? Because VS is saying it is not used

martincostello replied:
If only some TFMs use it, then you'll need to wrap it in an `#if` to stop the suggestion to remove it.

remcolam replied:
Yeah, it is needed, still strange that visual studio isn't recognizing it as being used

remcolam replied:
restored it

martincostello replied:
It's probably because you have a TFM that isn't the one that needs it selected as the current "view" (I don't know what the terminology for it is) you have of the code open in the document well.

remcolam commented:
BTW, should there be documentation in the readme.md about how to use this?

martincostello commented:
> BTW, should there be documentation in the readme.md about how to use this?

If you can open a new PR (to save more churn) once this is merged that adds a small note about it somewhere in the README that would be great 👍

remcolam commented:
> > BTW, should there be documentation in the readme.md about how to use this?
> 
> If you can open a new PR (to save more churn) once this is merged that adds a small note about it somewhere in the README that would be great 👍

Will do

martincostello commented:
Thanks again for the work you've put into this change!
